### ISSUE TITLE
is it possible to sign alias domains with dkim?

### ISSUE THREAD

**Issue by BaNzOiD** [modoboa/modoboa#1882]
# Impacted versions

* Modoboa: 1.14.0
* installer used: Yes
* Webserver: Nginx
* Distro Debian 10 Buster

# Question

I have configured dkim on the master domain but I have some users that send email from the alias domain connected to the master domain.  I would like those mails to be properly dkim signed.

In /etc/opendkim.com the data is taken from the dkim postgres table:

```bash
KeyTable		dsn:pgsql://opendkim:XXXXX@127.0.0.1/modoboa/table=dkim?keycol=id?datacol=domain_name,selector,private_key_path
SigningTable		dsn:pgsql://opendkim:XXXXX@127.0.0.1/modoboa/table=dkim?keycol=domain_name?datacol=id
```

Which is actually a view that gets its values from admin_domain table:

```bash
modoboa=# \d dkim
                             View "public.dkim"
      Column      |          Type          | Collation | Nullable | Default
------------------+------------------------+-----------+----------+---------
 id               | integer                |           |          |
 domain_name      | character varying(100) |           |          |
 private_key_path | character varying(254) |           |          |
 selector         | character varying(30)  |           |          |

modoboa=# \d+ public.dkim
                                         View "public.dkim"
      Column      |          Type          | Collation | Nullable | Default | Storage  | Description
------------------+------------------------+-----------+----------+---------+----------+-------------
 id               | integer                |           |          |         | plain    |
 domain_name      | character varying(100) |           |          |         | extended |
 private_key_path | character varying(254) |           |          |         | extended |
 selector         | character varying(30)  |           |          |         | extended |
View definition:
 SELECT admin_domain.id,
    admin_domain.name AS domain_name,
    admin_domain.dkim_private_key_path AS private_key_path,
    admin_domain.dkim_key_selector AS selector
   FROM admin_domain
  WHERE admin_domain.enable_dkim;

                                            Table "public.admin_domain"
        Column         |           Type           | Collation | Nullable |                 Default
-----------------------+--------------------------+-----------+----------+------------------------------------------
 id                    | integer                  |           | not null | nextval('admin_domain_id_seq'::regclass)
 name                  | character varying(100)   |           | not null |
 default_mailbox_quota | integer                  |           | not null |
 enabled               | boolean                  |           | not null |
 type                  | character varying(20)    |           | not null |
 creation              | timestamp with time zone |           | not null |
 last_modification     | timestamp with time zone |           | not null |
 enable_dns_checks     | boolean                  |           | not null |
 quota                 | integer                  |           | not null |
 transport_id          | integer                  |           |          |
 dkim_key_length       | integer                  |           |          |
 dkim_key_selector     | character varying(30)    |           | not null |
 dkim_private_key_path | character varying(254)   |           | not null |
 dkim_public_key       | text                     |           | not null |
 enable_dkim           | boolean                  |           | not null |
Indexes:
    "admin_domain_pkey" PRIMARY KEY, btree (id)
    "admin_domain_name_key" UNIQUE CONSTRAINT, btree (name)
    "admin_domain_transport_id_key" UNIQUE CONSTRAINT, btree (transport_id)
    "admin_domain_name_224678fe_like" btree (name varchar_pattern_ops)
Check constraints:
    "admin_domain_default_mailbox_quota_06dd964e_check" CHECK (default_mailbox_quota >= 0)
    "admin_domain_dkim_key_length_check" CHECK (dkim_key_length >= 0)
    "admin_domain_quota_check" CHECK (quota >= 0)
Foreign-key constraints:
    "admin_domain_transport_id_7e20103c_fk_transport_transport_id" FOREIGN KEY (transport_id) REFERENCES transport_transport(id) DEFERRABLE INITIALLY DEFERRED
Referenced by:
    TABLE "admin_alias" CONSTRAINT "admin_alias_domain_id_cf87ebf8_fk_admin_domain_id" FOREIGN KEY (domain_id) REFERENCES admin_domain(id) DEFERRABLE INITIALLY DEFERRED
    TABLE "admin_dnsblresult" CONSTRAINT "admin_dnsblresult_domain_id_eaf09b4b_fk_admin_domain_id" FOREIGN KEY (domain_id) REFERENCES admin_domain(id) DEFERRABLE INITIALLY DEFERRED
    TABLE "admin_domainalias" CONSTRAINT "admin_domainalias_target_id_99f9dbcd_fk_admin_domain_id" FOREIGN KEY (target_id) REFERENCES admin_domain(id) DEFERRABLE INITIALLY DEFERRED
    TABLE "admin_mailbox" CONSTRAINT "admin_mailbox_domain_id_401a6f1e_fk_admin_domain_id" FOREIGN KEY (domain_id) REFERENCES admin_domain(id) DEFERRABLE INITIALLY DEFERRED
    TABLE "admin_mxrecord" CONSTRAINT "admin_mxrecord_domain_id_b5400aa4_fk_admin_domain_id" FOREIGN KEY (domain_id) REFERENCES admin_domain(id) DEFERRABLE INITIALLY DEFERRED
    TABLE "dnstools_dnsrecord" CONSTRAINT "dnstools_dnsrecord_domain_id_251b0a52_fk_admin_domain_id" FOREIGN KEY (domain_id) REFERENCES admin_domain(id) DEFERRABLE INITIALLY DEFERRED
    TABLE "limits_domainobjectlimit" CONSTRAINT "limits_domainobjectlimit_domain_id_1b89a238_fk_admin_domain_id" FOREIGN KEY (domain_id) REFERENCES admin_domain(id) DEFERRABLE INITIALLY DEFERRED
    TABLE "radicale_sharedcalendar" CONSTRAINT "radicale_sharedcalendar_domain_id_f9efd3d8_fk_admin_domain_id" FOREIGN KEY (domain_id) REFERENCES admin_domain(id) DEFERRABLE INITIALLY DEFERRED
```

Is there a way to enable dkim signing for alias domains from the web interface or via api?
If it must be done by executing an insert via psql cli, can you point me in the right direction with an example? My db skills are not good enough yet :-(

Thanks in advance for your help!

**Comment by tonioo** [comment]
@BaNzOiD that's a good question! Currently, there is no way to sign messages from domain aliases. It would require to:
1. Generate DKIM keys for domain aliases, or use the one associated to the domain they point to
2. Update the SQL view (see https://github.com/modoboa/modoboa-installer/blob/master/modoboa_installer/scripts/files/opendkim/dkim_view_postgres.sql)
3. Maybe add an option on domain level to enable DKIM signing for aliases

**Comment by PatTheMav** [comment]
This gets a +1 from me, because I only noticed this due to a DMARC report from google telling me I failed the DKIM test (which it doesn't usually), but the obvious reason was that I sent an email from an alias domain to a gsuite-managed email address.

I'd be fine with using the same key for aliases (and adding them to my DNS), opendkim would "just" need the entries for the aliases then. Should that be possible to do "manually"?
